src/CurlUtil.cc

```diff
@@ -1039,6 +1050,81 @@ CurlWorker::Run() {
     m_op_map.clear();
 }
 
+bool
+CurlWorker::RefreshCacheToken() {
+    auto [success, contents] = RefreshCacheTokenStatic(m_token_file, m_logger);
+    if (success && !contents.empty()) {
+        m_cache_token = contents;
+    }
+    return success;
+}
+
+std::pair<bool, std::string>
+CurlWorker::RefreshCacheTokenStatic(const std::string &token_location, XrdCl::Log *log) {
+    if (token_location.empty()) {
+        return {true, ""};
+    }
+
+    std::string line;
+    std::ifstream fhandle;
+    fhandle.open(token_location);
+    if (!fhandle) {
+        log->Error(kLogXrdClPelican, "Cache token location is set (%s) but failed to open (worker PID %d): %s", token_location.c_str(), getthreadid(), strerror(errno));
+        return {false, ""};
+    }
+
+    std::string result;
+    while (std::getline(fhandle, line)) {
+        rtrim(line);
+        ltrim(line);
+        if (line.empty() || line[0] == '#') {continue;}
+
+        result = line;
+    }
+    if (!fhandle.eof() && fhandle.fail()) {
+        log->Error(kLogXrdClPelican, "Reading of token file (%s) failed: %s", token_location.c_str(), strerror(errno));
+        return {false, ""};
+    }
+    return {true, result};
+}
+
+bool
+CurlWorker::SetupCacheToken(CURL *curl) {
+    return SetupCacheTokenStatic(m_cache_token, curl, m_logger);
+}
+
+bool
+CurlWorker::SetupCacheTokenStatic(const std::string &token, CURL *curl, XrdCl::Log *log) {
+    if (!curl) {return false;}
+    if (token.empty()) {return true;}
+
+    char *url_char = nullptr;
+    CURLcode errnum;
+    if ((errnum = curl_easy_getinfo(curl, CURLINFO_EFFECTIVE_URL, &url_char)) != CURLE_OK) {
+        log->Error(kLogXrdClPelican, "Failed to get the CURL handle's current URL: %s", curl_easy_strerror(errnum));
+        return false;
+    }
+    if ((url_char == nullptr) || !url_char[0]) {
+        log->Error(kLogXrdClPelican, "Curl handle returned an empty URL");
+        return false;
+    }
+
+    std::string_view url{url_char};
+    auto has_query_string = url.find('?') != std::string::npos;
+    std::string final_url{url};
+    final_url += has_query_string ? "&" : "?";
+    final_url += "access_token=";
+    final_url += token;
```

Comment:
This type of URL parsing feels like it should be common enough (if not already, then eventually) to make a utility function.

Reply:
Agreed!

This has a lot of the characteristics of a generic URL parser -- but only targets extracting a a single parameter.  I think I'm going to keep it as-is for this PR but tackle things next time we need a "one-off".

Reply:
Agreed!

This has a lot of the characteristics of a generic URL parser -- but only targets extracting a a single parameter.  I think I'm going to keep it as-is for this PR but tackle things next time we need a "one-off".

Reply:
Agreed!

This has a lot of the characteristics of a generic URL parser -- but only targets extracting a a single parameter.  I think I'm going to keep it as-is for this PR but tackle things next time we need a "one-off".